### DISCUSSION TITLE
serialization  problem with "f32 as f64"

### DISCUSSION THREAD

**Discussion by youngsofun** [databendlabs/databend#5863] [General]
this happen in both seder_json::Value and our DataValue,  for a standalone value they use a Float64 to hold both f32 and f64 (e.g. serde_json::to_value() do this). for f32, use "as f64" to cast.

it is ok for calc.
but cause problems to serialization,  in this case:  it will  serialize 12.34 to 12.34000015258789

for our owner DataValue, we can solve it by skipping the DataValue and using the raw column data for serialization( serialize f32 directly). 
https://github.com/datafuselabs/databend/pull/5791/files

but for seder_json, we use seder_json::Value as AST a lot, one we use it,  we lost the info that it is originally f32.

since HTTP handler use nested JSON (by nested I mean a field is a json ast too ) for the result.
the diff of the two make it hard to write logistic test for both handlers.

----

the problem did not show up, because we use DataValue for serialization, so get the same result with serde_json

**Comment by youngsofun** [comment]
the code show

1.  it is a real problem for json
2.  the cause as `v32 as f64`, not ryn
 
```
// json  use ryn to format f32

fn p(v32: f32, v64: f64) {
    let vv64 = v32 as f64;
    let vv32 = vv64 as f32;
    let mut b = ryu::Buffer::new();
    let mut b2 = ryu::Buffer::new();
    let mut b3 = ryu::Buffer::new();
    let mut b4 = ryu::Buffer::new();
    println!("{} {} {} {}", v32, v64, vv32, vv64);
    println!(
        "ryu {} {} {} {}",
        b.format_finite::<f32>(v32),
        b2.format_finite::<f64>(v64),
        b4.format_finite::<f32>(vv32),
        b3.format_finite::<f64>(vv64),
    );

    // to_value() call serialize(),
    //   fn serialize_f32(self, value: f32) -> Result<Value> {
    //         self.serialize_f64(value as f64)
    //     }

    println!(
        "j {} {}",
        serde_json::to_value::<f32>(v32).unwrap().to_string(),
        serde_json::to_value::<f64>(v64).unwrap().to_string()
    );
}
fn main() {
    let s = "12.34";
    let v64 = s.parse::<f64>().unwrap();
    let v32 = v64 as f32;
    p(v32, v64)
}

```

output:

```
12.34 12.34 12.34 12.34000015258789
ryu 12.34 12.34 12.34 12.34000015258789
j 12.34000015258789 12.34
```

**Comment by youngsofun** [comment]
12.34000015258789 is bad but maybe we have to tolerate if for now if can not get a round serder_json::Value

@sundy-li

**Comment by sundy-li** [comment]
Clickhouse has settings to quoted some results.

```
    M(Bool, output_format_json_quote_64bit_integers, true, "Controls quoting of 64-bit integers in JSON output format.", 0) \
    M(Bool, output_format_json_quote_denormals, false, "Enables '+nan', '-nan', '+inf', '-inf' outputs in JSON output format.", 0) \
```

**Comment by youngsofun** [comment]
> Clickhouse has settings to quoted some results.
> 
> ```
>     M(Bool, output_format_json_quote_64bit_integers, true, "Controls quoting of 64-bit integers in JSON output format.", 0) \
>     M(Bool, output_format_json_quote_denormals, false, "Enables '+nan', '-nan', '+inf', '-inf' outputs in JSON output format.", 0) \
> ```

snowflake API use strings too
https://docs.snowflake.com/en/developer-guide/sql-api/handling-responses.html#getting-the-data-from-the-results

-----

propose we turn the HTTP handler to return JSON with a string for each field.  I will start with adding a flag to select which to use.

1.  not lose anything for JSON format
2. it will make the results of HTTP handler and MySQL   consistent  (MySQL use text protocal for Query)


-----

 for sqllogic test @ZeaLoVe 

field is turn to str in python, only need to remove some type checks here, I will do it
https://github.com/ZeaLoVe/databend/blob/44a3f4b141/tests/logictest/http_runner.py#L75

---

for front end @Chasen-Zhang 
I guess it is ok, since only need to show it

---

for HTTP handler unit test, I will use original version for now, and change to all-string a little later.

**Comment by youngsofun** [comment]
@sundy-li  done in https://github.com/datafuselabs/databend/pull/5791 thank you for advise！
